### ISSUE TITLE
Open html emails in browser via mailcap entry.

### ISSUE THREAD

**Issue by hsanson** [pazz/alot#1153]
Please ensure that the following information is filled before filing
--------------------------------------------------------------------

Please try to verify that the issue exists in the latest release or, even
better, the current master.

Please run alot again with the options `-d debug -l log`.

Software Versions
-----------------

Python version:2.7.12

Alot version (either a number or sha for master): 0.6.0dev

Observed behavior
-----------------

Cannot open html emails in a browser via mailcap entry.

Expected behavior
-----------------

In mutt when a message has both html and plain text parts it will render the html (via w3m). But it will also show both parts as attachments and pressing enter on the html part will open a browser with the html version of the email rendered as the sender intended it to be. This is extremely useful to print those receipts that are sent via email in html format.

This is the mailcap entry that makes this happen in mutt:

```
text/html; google-chrome %s & sleep 3; test=test -n "$DISPLAY"; needsterminal;
text/html; w3m -dump -o document_charset=%{charset} '%s'; nametemplate=%s.html; copiousoutput
```

It would be nice for alot to have the same or similar behavior. That is render the html as text using w3m or any html2text converter and open the html part in a browser when pressing enter in the html part (attachment).

Or if someone could add documentation in the wiki with an example on how to pipe the html part only and open it in a browser would suffice.

**Comment by dcbaker** [comment]
Alot does support using a mailcap file, the installation section of the rtd page has information on it (https://alot.readthedocs.io/en/latest/installation.html?highlight=mailcap), including exactly the w3m command you have in yours.

**Comment by hsanson** [comment]
@dcbaker sorry if my explanation was not clear. With the above mailcap alot shows the html emails in the terminal as expected. This works wonderfully and I have no issues with it.

But the first entry in the mailcap, the one that invokes an external browser, seems to be completely ignored by alot. In mutt the first entry allows me to press enter in the html part and get it displayed on a browser while the second entry displays the html email inline within the terminal.

**Comment by dcbaker** [comment]
Okay, so it sounds like what you want is the ability to describe multiple mailcap entries, and have alot use one like w3m to dump the mail, and then have a second mailcap entry that can be used to open it in an external program?

**Comment by hsanson** [comment]
Yes, that is what I am talking about.

I found the Mutt documentation explaining this behavior:

   http://www.mutt.org/doc/manual/#mailcap-search-order

This allows for quick reading of html emails online and at the same time allows for opening them in a browser for printing.

Other solution is to create a key binding using pipeto that outputs the html part to  a file and opens a brower with it but I have not been able to find a way to pipe only the html part of emails.

**Comment by pazz** [comment]
Hi!

I don't think there is a nice way to do this in alot at the moment: the
trouble is that alot does not allow to (easily) display or extract
parts of the MIME-tree of a mail.

This said, what you could do to get your feature would be to write (and
bind to a key) some hook code to be called in thread mode, which

1. gets the selected message:
   msg = ui.current_buffer.get_selected_message()
2. gets its email message
   eml = msg.get_email()
3. uses alot.db.utils.extract_body (see http://alot.readthedocs.io/en/latest/api/database.html#alot.db.utils.extract_body) to get extract all html parts:

   htmlstr = alot.db.utils.extract_body(eml, types='text/html')
4. writes this to a tempfile and calls your browser on it.

I hope this helps, also check out the contrib-hooks wiki for some
inspiration: https://github.com/pazz/alot/wiki/Contrib-Hooks
If you get this going, don't hesitate to drop your code in that wiki
as well :)

**Comment by hsanson** [comment]
@pazz thanks for the tips. That is enough to get me started.
Now I am having issues getting hooks to work for me (see #1156) but once I get that sorted I think I will get the functionality I need.

**Comment by hsanson** [comment]
@pazz I got this working based on your tips. The only problem I had is that the alot.db.utils.extract_body() method actually uses mailcap to convert the html to text if there is an entry for it. Fortunately I was able to use the same code used in that method to extract the raw html part.

Adding an additional flag to the extract_body() method to skip mailcap would simplify this hook a lot.

Added it to the Contrib-Hooks so others can benefit from it.
